crates/bitwarden-crypto/src/keys/symmetric_crypto_key.rs

```diff
@@ -1,83 +1,206 @@
-#[cfg_attr(test, derive(PartialEq))]
 pub enum SymmetricCryptoKey {
     Aes256CbcKey(Aes256CbcKey),
     Aes256CbcHmacKey(Aes256CbcHmacKey),
+    // always encode with cose
+    XChaCha20Poly1305Key(XChaCha20Poly1305Key),
 }
 
 impl SymmetricCryptoKey {
-    const KEY_LEN: usize = 32;
-    const MAC_LEN: usize = 32;
+    // enc type 0 old static format
+    const AES256_CBC_KEY_LEN: usize = 32;
+    // enc type 2 old static format
+    const AES256_CBC_HMAC_KEY_LEN: usize = 64;
+
+    /**
+     * Generate a new random AES256_CBC_HMAC [SymmetricCryptoKey]
+     */
+    pub fn generate() -> Self {
+        let mut rng = rand::thread_rng();
+        Self::generate_internal(&mut rng, false)
+    }
+
+    /**
+     * Generate a new random XChaCha20Poly1305 [SymmetricCryptoKey]
+     */
+    pub fn generate_xchacha20() -> Self {
+        let mut rng = rand::thread_rng();
+        Self::generate_internal(&mut rng, true)
+    }
 
     /// Generate a new random [SymmetricCryptoKey]
-    pub fn generate(mut rng: impl rand::RngCore) -> Self {
-        let mut enc_key = Box::pin(GenericArray::<u8, U32>::default());
-        let mut mac_key = Box::pin(GenericArray::<u8, U32>::default());
+    /// @param rng: A random number generator
+    /// @param xchacha: If true, generate an XChaCha20Poly1305 key, otherwise generate an
+    /// AES256_CBC_HMAC key
+    pub(crate) fn generate_internal(mut rng: impl rand::RngCore, xchacha20: bool) -> Self {
+        if !xchacha20 {
+            let mut enc_key = Box::pin(GenericArray::<u8, U32>::default());
+            let mut mac_key = Box::pin(GenericArray::<u8, U32>::default());
 
-        rng.fill(enc_key.as_mut_slice());
-        rng.fill(mac_key.as_mut_slice());
+            rng.fill(enc_key.as_mut_slice());
+            rng.fill(mac_key.as_mut_slice());
 
-        SymmetricCryptoKey::Aes256CbcHmacKey(Aes256CbcHmacKey { enc_key, mac_key })
+            SymmetricCryptoKey::Aes256CbcHmacKey(Aes256CbcHmacKey { enc_key, mac_key })
+        } else {
+            let mut enc_key = Box::pin(GenericArray::<u8, U32>::default());
+            rng.fill(enc_key.as_mut_slice());
+            SymmetricCryptoKey::XChaCha20Poly1305Key(XChaCha20Poly1305Key {
+                enc_key,
+                key_id: *KeyId::generate().as_bytes(),
+            })
+        }
     }
 
-    fn total_len(&self) -> usize {
+    /// Encodes the key to a byte array representation, that is separated by size.
+    /// `SymmetricCryptoKey::Aes256CbcHmacKey` and `SymmetricCryptoKey::Aes256CbcKey` are
+    /// encoded as 64 and 32 bytes respectively. `SymmetricCryptoKey::XChaCha20Poly1305Key`
+    /// is encoded as at least 65 bytes, by using padding defined in `pad_key`.
+    ///
+    /// This can be used for storage and transmission in the old byte array format.
+    /// When the wrapping key is a COSE key, and the wrapped key is a COSE key, then this should
+    /// not use the byte representation but instead use the COSE key representation.
```

Comment:
Please link to the correct method to use ([rustdocs book](https://doc.rust-lang.org/rustdoc/write-documentation/linking-to-items-by-name.html))

Reply:
It seems it is not allowed to link private methods in public methods.

Reply:
I removed the reference to the method entirely and rewrote it to say "using padding".